Reading/Writing to case class from a parsed csv

tbrown1979 opened the pull request:
#### What this PR does

* Read a ParsedCSV to a case class
* Write a Seq[CaseClass] to a ParsedCSV
* Just a matter of adding an implicit in the companion object of your case class to read/write 
* Hold on to any errors that are encountered in this process (i.e. parsing errors, read/writing errors)
* Allows for custom types. You can define your own ColumnFormats to convert a column(string) to your own type.

tbrown1979 commented:
@adamdecaf  So...here's sort of where I went with this. Let me know what you think.

adamdecaf commented:
@tbrown1979 I've been thinking and it might make more sense to leave this parsing left to the client of this library. There's lots of edge cases around leading zero's in numbers, spaces or hyphens that might make the parsing in this library to complex.

adamdecaf commented:
@tbrown1979 Yea, this makes sense. It also allows the chance to use custom types. (An Email Reader for example.)

tbrown1979 commented:
@adamdecaf This should be complete.

adamdecaf commented:
Looks pretty good! :+1:

zmccoy commented:
:+1: :+1:

tbrown1979 commented:
Went ahead and updated the readme as well

adamdecaf commented on src/main/scala/CSVParser.scala:
```diff
@@ -17,6 +17,17 @@ object CSVParser {
+    val validatedParsedCSV = apply(enum, optional).parseCSVFile(csv)
+    validatedParsedCSV.fold(errors => errors.failure[Seq[T]], parsedCSV => reader.read(parsedCSV))
+  }
+
```
Does it work if we curry the `enum` and `optional` params? I'm wondering if we can figure out a way to not have to specify the `Headers.type` explicitly.

tbrown1979 replied:
Yeah, I wasn't too happy with that. I tried thinking of some way where I didn't have to explicitly say `Headers.type` but I couldn't find a solution. I just tried currying but it still tells me 
`wrong number of type parameters for method parseCSVTo`. 

I'd really prefer for the call to work like this: 
```scala
CSVParser.parseCSVTo[Person]("csv as string", Headers)
```
Just not sure how to get it there.

adamdecaf replied:
@tbrown1979 how does https://github.com/Banno/banno-csv/pull/16 look? I can explain how it works under the hood if you want.

tbrown1979 replied:
:+1: I like it. I was trying to do something like that with functions but kept hitting the same barrier as before. This is a nice way to fix the problem.

adamdecaf replied:
@tbrown1979 Yea, it's a problem with scala and how it infers types for parameter lists...

tbrown1979 commented:
Not quite ready to be merged yet. Was missing tests for fromCSV and toCSV. Gonna get those done, which is going to require some changes.

tbrown1979 commented on src/main/scala/CSVParser.scala:
```diff
@@ -17,6 +17,25 @@ object CSVParser {
+      reader.read(parsedCSV)
+  }
+
+  def toCSV[T, E <: Enumeration](from: Seq[T])(implicit writer: CSVWriter[E#Value, T]): ParsedCSV[E#Value] =
```
@adamdecaf I could be mistaken, but this really appears to be a spot where you're going to have to say `Headers.type`. I don't see any way around it on this. 
Above, I had to do basically the same thing for the `fromCSV` function that you did for the `parseCSVTo`

adamdecaf replied:
We could just do that for this `toCSV` call also.

tbrown1979 replied:
It doesn't take in an Enumeration/ParsedCSV like the others and so I'm not really sure how it would work. We really only use the type of the Enumeration for the `toCSV` function

adamdecaf replied:
I think we should just use the builder pattern here and add `.toCSV[E]` on the end. Make sense?

tbrown1979 replied:
Yep, that makes sense. Got it changed!

tbrown1979 commented on src/test/scala/CSVParserSpec.scala:
```diff
@@ -6,231 +6,306 @@ import org.specs2.scalaz.ValidationMatchers
 object CSVParserSpec extends Specification with ValidationMatchers {
+  "CSVParser" should {
+    "parse from a parsedCSV to a case class" in new context {
+      object Decimal {
```
Compiler keeps complaining if I don't have the case class and the object directly in the test. If I try to put these in the `context` it just keeps erroring with: 

```scala
[error] could not find implicit value for parameter reader: com.banno.csv.CSVReader[this.DecimalHeaders.Value,this.Decimal]
[error]       CSVParser(parsedDecimal).fromCSV[Decimal] must beSuccessful.like {
[error]
``` 
Not really sure what's going on here

adamdecaf replied:
@tbrown1979 That's probably related to the order in which values are typed and put into the scope for implicit search. (Implicit search a) doesn't infer types nor b) compile / see declarations after the call point of the block it's in.)

tbrown1979 replied:
Alrighty, that makes sense.

rubbish commented:
:+1: btw someday I'm going to rewrite how jsonz does these product formats to use shapeless.

tbrown1979 commented:
Alrighty, I changed a bunch of stuff. 

* Enumerations are no longer used as headers. Instead, a `Seq[String]` is used. This kinda cleans up having to use that `Field <: Enumeration#Value` guy all over the place. Note: In the product formats file I just did `type E = String`. Might seem silly, but if the type changes again we only have to change that one line.

* The productFormats now convert ParsedCSVRows to/from a single case class. The `toCSV` | `fromCSV` | `parseCSVTo` functions now handle the reading/writing of ParsedCSVs and Seq[CaseClass]. This is ultimately more flexible. To alter how a ParsedCSV gets read to a case class we can just change the single fromCSV function. 

* I removed the `MissingColumn` error. Basically, if a column value is empty in a CSV that's okay. If a column doesn't appear on a line in the CSV then we have a problem and we throw the `TooFewColumns` error. This means that `MissingColumn` is just never used now, so I removed it.

InfernoJJ commented on src/main/scala/ProductFormats.scala:
```diff
@@ -0,0 +1,226 @@
+trait CSVFormat[T] extends CSVReader[T] with CSVWriter[T]
+
+object CSVProductFormats {
+  type E = String
```
Can we change E to a more clean-code name.

tbrown1979 replied:
Changed to `Field`.

adamdecaf commented:
@tbrown1979 I'll read this over today/tomorrow.

adamdecaf commented on README.md:
```diff
@@ -9,21 +9,46 @@
+  val email = "Email"
+  val values = Seq(firstName, lastName, email)
 }
 
```
There is a `dev values` on any `scala.Enumeration` already. 

```
scala> object Foo extends Enumeration
defined object Foo

scala> Foo.values
res0: Foo.ValueSet = Foo.ValueSet()

scala> Foo.values.toList
res2: List[Foo.Value] = List()
```

adamdecaf replied:
Oh wait, these aren't enumerations anymore.

adamdecaf commented:
:+1:

zmccoy commented on src/main/scala/ProductFormats.scala:
```diff
@@ -0,0 +1,226 @@
+  }
+
+  def productFormat8[T1, T2, T3, T4, T5, T6, T7, T8, R](f1: Field, f2: Field, f3: Field, f4: Field, f5: Field, f6: Field, f7: Field, f8: Field)(apply: (T1, T2, T3, T4, T5, T6, T7, T8) => R)(unapply: R => Option[(T1, T2, T3, T4, T5, T6, T7, T8)])(implicit r1: CSVColumnFormat[T1], r2: CSVColumnFormat[T2], r3: CSVColumnFormat[T3], r4: CSVColumnFormat[T4], r5: CSVColumnFormat[T5], r6: CSVColumnFormat ...
+    def read(row: ParsedCSVRow) = readColumn[T8](row, f8, r8) <*> (readColumn[T7](row, f7, r7) <*> (readColumn[T6](row, f6, r6) <*> (readColumn[T5](row, f5, r5) <*> (readColumn[T4](row, f4, r4) <*> (readColumn[T3](row, f3, r3) <*> (readColumn[T2](row, f2, r2) <*> (readColumn[T1](row, f1, r1) map apply.curried)))))))
```
TIL Applicative Builder syntax only goes up to 12.

iliffwes commented:
:+1:
